### ISSUE TITLE
Lint: High cyclomatic complexity

### ISSUE THREAD

**Issue by killercup** [Manishearth/rust-clippy#418]
[Cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) can be used to see how "complex" a bit of source code is (it counts independent paths).

Clippy could have an (allow by default) lint, which warns the developer when a function's complexity is above a certain value and suggests refactoring it into smaller, independent functions or methods.

**Comment by Manishearth** [comment]
http://manishearth.github.io/rust-internals-docs/rustc/middle/cfg/struct.CFG.html#method.new sounds useful here

**Comment by llogiq** [comment]
This would be a simple use case for the ControlFlowVisitor I suggested recently.

**Comment by oli-obk** [comment]
is there a way to parametrize lints? like `#[warn(cyclomatic_complexity(20))]` and `#[error(cyclomatic_complexity(100))]`

It would be trivial to add a syntax extension so you'd call

cyclomatic_complexity_error!(100);
cyclomatic_complexity_warn!(20);

but that's kindof weird.

**Comment by oli-obk** [comment]
Also a slight problem with the cyclomatic complexity is that matching is considered bad. Every match arm increases the cc by one. That'll make most Expr-processing functions in the Rust compiler have at least a cc of 30, which is way too high for most normal functions. Also `if let` statements would increase the cc by two instead of one like `if` statements. Special-casing matches to only increase the cc by the maximum cc of its arms might be the way to go here.

**Comment by Manishearth** [comment]
The Rust compiler is not an example of idiomatic Rust code.

And there will always be edge cases.

**Comment by oli-obk** [comment]
What do you see as an edge case here?

**Comment by Manishearth** [comment]
Rustc and clippy having high cyclomatic complexity. It's because of the large number of non-box pointers (`P`) being used in the AST, along with the `Spanned` business, which makes destructuring hard. Most of these cases are really just one match with a complex pattern.

But yes, matches are a more common form of branching which we should tweak against.

**Comment by oli-obk** [comment]
Ah yes, I did not mean hierarchical matches, these are a sign of complexity and should be reported as such.

Good, I'll tweak the cc computation and post the results here.

**Comment by Manishearth** [comment]
Note that the lint could be allow by default, so we can land it first, let it be used by people, and tweak it accordingly.

**Comment by oli-obk** [comment]
About the tweaking, do you know if something like `#[error(cyclomatic_complexity(100))]` to pass arguments to the lint is possible? Or what is your opinion on such a feature?

**Comment by Manishearth** [comment]
Not possible as part of the lint specifier itself. You can leave a crate attribute `#![cyclo(100)]` and have the lint catch that, or perhaps use http://manishearth.github.io/rust-internals-docs/rustc/lint/trait.LateLintPass.html#method.enter_lint_attrs

**Comment by llogiq** [comment]
I had a pre-RFC on rust-internals, but I'm totally swamped with work.

**Comment by flying-sheep** [comment]
Macros still count towards the limit. e.g. I have

```rust
#[test]
fn two_targets() {
    parses_to! {
        parser: RstParser,
        input: "\
.. _a: http://example.com
.. _`b_`: https://example.org
",
        rule: Rule::document,
        tokens: [
            target(0, 26, [
                target_name_uq(4, 5),
                link_target(7, 25),
            ]),
            target(26, 56, [
                target_name_qu(31, 33),
                link_target(36, 55),
            ]),
        ]
    };
}
```

which apparently has a cc of 29. (macro from `pest`)

**Comment by oli-obk** [comment]
We should probably just skip `#[test]` functions or skip the lint in general if the `test` feature is active

**Comment by flying-sheep** [comment]
I’m not sure about that: tests can get too complex as well.

But this test only contains a single macro invocation, which is simple enough.

**Comment by flying-sheep** [comment]
Hmm, you’re right. I think disabling on a case-by-case basis doesn’t hurt too much, and I can always simplify the macro by extracting some code to a function.

**Comment by oli-obk** [comment]
You can also try adding `#![allow(clippy::cyclomatic_complexity)]` to the macro ;) maybe all functions using the macro will stop reporting the lint then.
